### PR TITLE
Feature 567: Controllers and actions refactor

### PR DESCRIPTION
**Pull request by dbengsch**
See #567 for initial purpose.

Follows a list of all controllers and actions in this PR (except `memoryStorage` that didn't change at all).
Associated HTTP routes have changed as well, please consult the [API documentation](http://kuzzle.io/api-reference/).

## Controller `auth`

- checkToken: No change
- getCurrentUser: No change
- getMyRights: No change
- logout: No change
- login: No change
- updateSelf: No change

## Controller `bulk`

- import: No change

## Controller `collection`

- create: Previously (write/createCollection)
- deleteSpecifications: Previously (admin/deleteSpecifications)
- exists: Previously (read/collectionExists)
- getMapping: Previously (admin/getMapping)
- getSpecifications: Previously (admin/getSpecifications)
- getUserMapping: Previously (admin/getUserMapping)
- searchSpecifications: Previously (admin/searchSpecifications)
- list: Previously (read/listCollections)
- truncate: Previously (admin/truncateCollection)
- updateMapping: Previously (admin/updateMapping)
- updateSpecifications: Previously (admin/updateSpecifications)
- updateUserMapping: Previously (admin/updateUserMapping)
- validateSpecifications: Previously (admin/validateSpecifications)

## Controller `document`

- count: Previously (read/count)
- create: Previously (write/create)
- createOrReplace: Previously (write/createOrReplace)
- delete: Previously (write/delete)
- deleteByQuery: Previously (write/deleteByQuery)
- get: Previously (read/get)
- replace: Previously (write/replace)
- scroll: Previously (read/scroll)
- search: Previously (read/search)
- update: Previously (write/update)
- validate: Previously (write/validate)

## Controller `index`

- create: Previously (admin/createIndex)
- delete: Previously (admin/deleteIndex)
- exists: Previously (read/indexExists)
- getAutoRefresh: Previously (admin/getAutoRefresh)
- list: Previously (read/listIndexes)
- mDelete: Previously (admin/deleteIndexes)
- refresh: Previously (admin/refreshIndex)
- refreshInternal: Previously (admin/refreshInternalIndex)
- setAutoRefresh: Previously (admin/setAutRefresh)

## Controller `realtime`

- count: Previously (subscribe/count)
- join: Previously (subscribe/join)
- list: Previously (subscribe/list)
- publish: Previously (write/publish)
- subscribe: Previously (subscribe/on)
- unsubscribe: Previously (subscribe/off)
- validate: Previously (write/validate)

## Controller `security`

- getRole: No change
- mGetRoles: No change
- searchRoles: No change
- createOrReplaceRole: No change
- createRole: No change
- deleteRole: No change
- getProfile: No change
- mGetProfiles: No change
- createOrReplaceProfile: No change
- createProfile: No change
- deleteProfile: No change
- searchProfiles: No change
- getUser: No change
- getProfileRights: No change
- getUserRights: No change
- searchUsers: No change
- deleteUser: No change
- createUser: No change
- createRestrictedUser: No change
- updateUser: No change
- updateProfile: No change
- updateRole: No change
- createOrReplaceUser: No change
- createFirstAdmin: Previously (admin/createFirstAdmin)

## Controller `server`

- adminExists: Previously (admin/adminExists)
- getAllStats: Previously (admin/getAllStats)
- getLastStats: Previously (admin/getLastStats)
- getStats: Previously (admin/getStats)
- getConfig: Previously (admin/getConfig)
- now: Previously (read/now)
- info: Previously (read/serverInfo)

### CODE REVIEW

File: default.config.js [hunk 1]
```diff
@@ -82,7 +82,7 @@ module.exports = {
                 updateSelf: true
               }
             },
-            read: {
+            server: {
               actions: {
                 serverInfo: true
```
Comment (ballinette) [inline default.config.js]: action name should now be "info" instead of "serverInfo" ?

File: default.config.js [hunk 2]
```diff
@@ -99,7 +99,7 @@ module.exports = {
                 login: true
               }
             },
-            read: {
+            server: {
               actions: {
                 serverInfo: true
```
Comment (ballinette) [inline default.config.js]: same as above

File: lib/api/controllers/documentController.js [hunk 3]
```diff
@@ -1,32 +1,77 @@
...
+function DocumentController (kuzzle) {
   var engine = kuzzle.services.list.storageEngine;
 
+
+  /**
+   * @param {Request} request
+   * @returns {Promise<Object>}
+   */
+  this.search = function documentSearch (request) {
+    assertHasIndexAndCollection(request, 'search');
```
Comment (ballinette) [inline lib/api/controllers/documentController.js]: as you changed "action" parameter above in other controllers, should we also replace `'search'` by `'document::search'` ?

**Comment by dbengsch** [comment]
@ballinette Done

File: lib/config/httpRoutes.js [hunk 4]
```diff
@@ -1,125 +1,40 @@
...
-            requestId: {type: 'object', description: 'The Request ID'},
-            result: {type: 'null', description: 'The result, here it is null'},
-            status: {type: 'integer', description: 'The status code, here 404'}
-          }
-        }
-      }
-    }
-  }},
+  {verb: 'get', url: '/roles/:_id', controller: 'security', action: 'getRole'},
+  {verb: 'get', url: '/users/:_id', controller: 'security', action: 'getUser'},
```
Comment (ballinette) [inline lib/config/httpRoutes.js]: Why did you remove the swagger infos ?

Comment (dbengsch) [inline lib/config/httpRoutes.js]: I don't see the point to keep 2 route with info over 150 without it. But in fact, I don't really matter.

**Comment by ballinette** [comment]
don't we have http routes for `realtime/count` and `realtime/validate` actions ?

File: lib/api/controllers/collectionController.js [hunk 5]
```diff
@@ -0,0 +1,396 @@
...
+  };
+
+  /**
+   * Get the collection validation specifications
+   *
+   * @param {Request} request
+   * @returns {Promise<Object>}
+   */
+  this.getSpecifications = function collectionGetSpecifications (request) {
+    return internalEngine.get('validations', `${request.input.resource.index}#${request.input.resource.collection}`)
```
Comment (dbengsch) [inline lib/api/controllers/collectionController.js]: done

File: lib/api/controllers/serverController.js [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+   * Returns the Kuzzle configuration
+   *
+   * @returns {Promise<Object>}
+   */
+  this.getConfig = function serverGetConfig () {
+    var
+      pluginsConfig = {},
+      promises = [];
+
+    Object.keys(kuzzle.pluginsManager.plugins).forEach(id => {
```
Comment (dbengsch) [inline lib/api/controllers/serverController.js]: done

**Comment by dbengsch** [comment]
@ballinette: We can add them in the future but for now we never had
